CORE-13747: Add RedpandaRoles to Whole Cluster Recovery

nguyen-andrew opened the pull request:
## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [x] v24.3.x

## Release Notes

### Bug Fixes

* Adding RedpandaRoles as part of whole cluster recovery. Prior to this change, roles were saved in the snapshot but not restored as part of the recovery process. This change adds that missing logic and extends the tests to validate the flow.

nguyen-andrew committed:
cluster: Adding roles from whole cluster recovery snapshot.

Updating controller_snapshot_reconciler::get_actions to extract
role actions from security snapshot, and then updating the
cluster_recovery_backend to apply the role actions to the
security_frontend. This role recovery will be bundled in with
the ACL recovery stage in order to avoid adding a new recovery
stage enum that would prevent backportablity.

nguyen-andrew committed:
cloud_metadata: Extending WCR unit tests for role recovery.

Extending the existing snapshot reconciliation and cluster recovery
unit tests to exercise role recovery.

nguyen-andrew committed:
rptest: Extend controller snapshot recover test for roles

Extending the test_basic_controller_snapshot_restore ducktape test
to validate the restoration of roles.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74756-->
<!--61cbdfc3-a1db-463b-9169-952aabd230a6-->
<!--title-->
## Retry command for [Build#74756](https://buildkite.com/redpanda/redpanda/builds/74756)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b61969ae-8e6c-4734-853e-c1d636b2cc48-->
<!--TEST RESULTS-->
<!--ef58b4f6-d575-4163-befc-e8b5100cbc6a-->
<!--TEST RESULTS-->
<!--5e307459-3649-440d-b43c-25ec7a83b1bb-->
<!--TEST RESULTS-->
<!--67df45c2-4ea1-46fb-abe3-79b926d1b63a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74756">build#74756</a></summary>


| test_class                          | test_method                | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                    | test_history                                                                                                                                                              |
|:------------------------------------|:---------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionMovementTest               | test_empty                 | {"num_to_upgrade": 0}       | integration | https://buildkite.com/redpanda/redpanda/builds/74756#019a0ed3-3204-49b8-a7bb-2011216ad8e6 | FLAKY         | 16/21    | upstream reliability is '99.3006993006993'. current run reliability is '76.19047619047619'. drift is 23.11022 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                            |
| PartitionMovementTest               | test_static                | {"num_to_upgrade": 0}       | integration | https://buildkite.com/redpanda/redpanda/builds/74756#019a0ecc-8b71-4249-bd15-91a6336385d3 | FLAKY         | 18/21    | upstream reliability is '99.06976744186046'. current run reliability is '85.71428571428571'. drift is 13.35548 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                           |
| ShadowLinkingRandomOpsTest          | test_node_operations       | {"failures": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/74756#019a0ed3-3209-48a6-a085-d99d1dc96392 | FLAKY         | 4/21     | upstream reliability is '36.891679748822604'. current run reliability is '19.047619047619047'. drift is 17.84406 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| ShadowLinkingRandomOpsTest          | test_node_operations       | {"failures": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/74756#019a0ecc-8b75-4a21-9baf-4f904be6e0f3 | FLAKY         | 1/21     | upstream reliability is '82.89473684210526'. current run reliability is '4.761904761904762'. drift is 78.13283 and the allowed drift is set to 50. The test should FAIL   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| ShadowLinkingRandomOpsTest          | test_node_operations       | {"failures": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/74756#019a0ed3-320a-4d7f-a59b-c684c0a8e958 | FLAKY         | 1/21     | upstream reliability is '32.72251308900523'. current run reliability is '4.761904761904762'. drift is 27.96061 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| WriteCachingFailureInjectionE2ETest | test_crash_all             | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74756#019a0ed3-3207-418a-863c-75d8f3346d2d | FLAKY         | 19/21    | upstream reliability is '81.67701863354037'. current run reliability is '90.47619047619048'. drift is -8.79917 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all          |
| WriteCachingFailureInjectionTest    | test_unavoidable_data_loss | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/74756#019a0ed3-320e-4076-8c1d-a034834c88e2 | FLAKY         | 20/21    | upstream reliability is '92.29885057471265'. current run reliability is '95.23809523809523'. drift is -2.93924 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74856">build#74856</a></summary>


| test_class                     | test_method                                             | test_arguments                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                   |
|:-------------------------------|:--------------------------------------------------------|:---------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkUpdateBrokersTests   | test_update_brokers                                     | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a135f-1f5f-47ed-a540-283543727f8c | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkUpdateBrokersTests&test_method=test_update_brokers                                 |
| DatalakeCustomPartitioningTest | test_sticky_default                                     | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a135f-1f5d-486f-aaee-8209c5a26d23 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeCustomPartitioningTest&test_method=test_sticky_default                               |
| NodesDecommissioningTest       | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}              | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a135f-1f5d-486f-aaee-8209c5a26d23 | FLAKY         | 20/21    | upstream reliability is '98.20971867007673'. current run reliability is '95.23809523809523'. drift is 2.97162 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| PartitionMovementTest          | test_empty                                              | {"num_to_upgrade": 0}                                    | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a1355-533b-42fa-9800-e9e6d1aca007 | FLAKY         | 19/21    | upstream reliability is '95.87813620071685'. current run reliability is '90.47619047619048'. drift is 5.40195 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                                                 |
| PartitionMovementTest          | test_static                                             | {"num_to_upgrade": 0}                                    | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a135f-1f5b-457e-9549-ea58ec0292ea | FLAKY         | 17/21    | upstream reliability is '91.92245557350566'. current run reliability is '80.95238095238095'. drift is 10.97007 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                                                |
| ShadowLinkingRandomOpsTest     | test_node_operations                                    | {"failures": false}                                      | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a1355-533f-4ada-bb26-4d8404ac91c6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest     | test_node_operations                                    | {"failures": false}                                      | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a135f-1f5d-486f-aaee-8209c5a26d23 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest     | test_node_operations                                    | {"failures": true}                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a1355-5340-4cee-a359-8c5af3daa854 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest     | test_node_operations                                    | {"failures": true}                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a135f-1f5f-47ed-a540-283543727f8c | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| TLSMetricsTest                 | test_labels                                             | null                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a1355-533e-48ee-a311-70508cfb0236 | FLAKY         | 7/21     | upstream reliability is '79.92277992277992'. current run reliability is '33.33333333333333'. drift is 46.58945 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSMetricsTest&test_method=test_labels                                                       |
| TLSVersionTestECDSA            | test_change_version                                     | {"version": 3}                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74856#019a135f-1f5d-486f-aaee-8209c5a26d23 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSVersionTestECDSA&test_method=test_change_version                                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74914">build#74914</a></summary>


| test_class               | test_method                                             | test_arguments                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                                   |
|:-------------------------|:--------------------------------------------------------|:---------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74914#019a16a4-74f0-4243-b7c2-34b4319ed38b | FLAKY         | 19/21    | upstream reliability is '98.1651376146789'. current run reliability is '90.47619047619048'. drift is 7.68895 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| PartitionMovementTest    | test_empty                                              | {"num_to_upgrade": 0}                        | integration | https://buildkite.com/redpanda/redpanda/builds/74914#019a16f4-4103-442b-a732-5033591eb461 | FLAKY         | 18/21    | upstream reliability is '94.0809968847352'. current run reliability is '85.71428571428571'. drift is 8.36671 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                                                 |
| RapidTopicRecreateTest   | test_topic_rapid_recreation                             | null                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74914#019a16f4-4103-442b-a732-5033591eb461 | FLAKY         | 20/21    | upstream reliability is '98.70370370370371'. current run reliability is '95.23809523809523'. drift is 3.46561 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RapidTopicRecreateTest&test_method=test_topic_rapid_recreation                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75123">build#75123</a></summary>


| test_class                        | test_method                                             | test_arguments                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                                   |
|:----------------------------------|:--------------------------------------------------------|:--------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade              | {"test_case_name": "Mixed aborts and commits"}    | integration | https://buildkite.com/redpanda/redpanda/builds/75123#019a2b98-26be-4757-8b7d-fa856e17efcb | FLAKY         | 19/21    | upstream reliability is '95.45454545454545'. current run reliability is '90.47619047619048'. drift is 4.97835 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade     |
| NodesDecommissioningTest          | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": false}     | integration | https://buildkite.com/redpanda/redpanda/builds/75123#019a2b98-26c8-4950-83bf-236afc01d383 | FLAKY         | 19/21    | upstream reliability is '98.85057471264368'. current run reliability is '90.47619047619048'. drift is 8.37438 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| RedpandaNodeOperationsSmokeTest   | test_node_ops_smoke_test                                | {"cloud_storage_type": 1, "mixed_versions": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75123#019a2b98-26bf-4f86-b878-bf3261350265 | FLAKY         | 20/21    | upstream reliability is '99.84871406959152'. current run reliability is '95.23809523809523'. drift is 4.61062 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test                         |
| TxAtomicProduceConsumeTest        | test_basic_tx_consumer_transform_produce                | {"with_failures": true}                           | integration | https://buildkite.com/redpanda/redpanda/builds/75123#019a2b98-26c5-4e05-8a9d-7440c39f8300 | FLAKY         | 18/19    | upstream reliability is '99.17184265010351'. current run reliability is '94.73684210526315'. drift is 4.435 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce              |
</details>
<!--/body-->

pgellert commented on tests/rptest/tests/cluster_recovery_test.py:
```diff
@@ -119,6 +133,19 @@ def test_basic_controller_snapshot_restore(self):
+        assert len(self.redpanda._admin.list_users()) == 0, (
+            "Expected no users before recovery"
+        )
+        assert len(rpk.acl_list().splitlines()) == 1, "Expected no ACLs before recovery"
```
Is this `== 1` because this prints the rpk output with the header?

nguyen-andrew replied:
Yes, exactly. I can add a comment to clarify

pgellert commented on tests/rptest/tests/cluster_recovery_test.py:
```diff
@@ -148,6 +175,31 @@ def cluster_recovery_complete():
+            res = rpk.describe_role(role_name)
+            restored_role_members = Counter(
+                RoleMember.User(member["name"])
+                for member in res.get("members", [])
+                if member["principal_type"] == RoleMember.PrincipalType.USER.value
```
Why do we need to filter for User principals here?

nguyen-andrew replied:
Right now it seems like User is the only principal type available. But I added this filter in case more types are added in the future - I'd like this to fail to draw attention to it so that the additional types can be handled.

pgellert commented on tests/rptest/tests/cluster_recovery_test.py:
```diff
@@ -148,6 +175,31 @@ def cluster_recovery_complete():
+                if member["principal_type"] == RoleMember.PrincipalType.USER.value
+            )
+
+            assert restored_role_members == members, (
```
TIL what a `Counter` collection is, thanks!

I think this could just be a `set` though because the same member shouldn't be associated with the same role multiple times.

nguyen-andrew replied:
Yea I originally started with a set, but I figured since we're just iterating through a list of string members returned from rpk, that list of members/strings technically _could_ contain anything (e.g. a bug causing duplicate members), so might as well have the test be able to catch that

pgellert replied:
I think I would disagree with "might as well have the test be able to catch that" because it makes the test more complex, less easy to follow.

I would lean towards making the test simpler, because using a Counter here makes me think that it is part of the expected behaviour that you would need to use a Counter here. So I would generally optimize for keeping the test simpler > handling unexpected cases. If I wanted to add a test for unexpected cases, I would make that a separate test + assertion, either here or in some more role-specific test suite.

But that's a nit for this case, explanation of my thought process in general.

pgellert commented on src/v/cluster/types.h:
```diff
@@ -1913,6 +1913,8 @@ enum class recovery_stage : int8_t {
     recovered_tx_coordinator = 10,
 
+    recovered_roles = 11,
+
```
On the slack thread there was some discussion about this being backportable as long as we don't introduce a new `recovery_stage` enum value. Do we plan to implement that (as part of this PR or a separate PR) or have we decided to not backport this fix? cc @michael-redpanda

michael-redpanda replied:
It would be nice to be able to backport this, and yes I thinka dding a new recovery stage enum entry makes this non backportable.  @andrwng would it be ok to just stick role recovery in the same stage as ACL recovery?  I think this makes the change backportable and so we can recover roles in v25.1 and v25.2 RP clusters (and maybe v24.3 if the backport works).

andrwng replied:
> would it be ok to just stick role recovery in the same stage as ACL recovery?

Yeah doing this as a part of ACL recovery is totally fine with me. We can always update it on dev later too.

pgellert commented on src/v/cluster/cluster_recovery_reconciler.cc:
```diff
@@ -104,6 +104,16 @@ controller_snapshot_reconciler::get_actions(
     }
 
+    if (cur_stage < recovery_stage::recovered_roles) {
+        const auto& snap_roles = snap.security.roles;
```
From the logic of the user recovery and the comment in the ACL recovery stage it looks like the intent is to only recover users/acls/roles if they don't exist. According to the comment the ACL staged, this is only skipped there because it doesn't make a difference due to idempotency. So I think we should feed in the role store here and check if the role exists before potentially overwriting its members.

nguyen-andrew replied:
Good point, will update

pgellert commented on src/v/cluster/cluster_recovery_reconciler.cc:
```diff
@@ -104,6 +104,16 @@ controller_snapshot_reconciler::get_actions(
         }
     }
 
+    if (cur_stage < recovery_stage::recovered_roles) {
```
If I am following the code correctly, this stage will actually run last because the ordering is based on the enum numbering. I was wondering if that's safe, but I think I've convinced myself that it is because you can create ACLs upfront without the role existing and the ducktape test also shows that we can recover roles + role ACLs.

michael-redpanda replied:
yeah you can definitely create ACLs against non-existant roles (just like you could create ACLs against non existing SCRAM credentials).

andrwng replied:
I don't feel strongly about it, but if we want to recover security metadata all together, we could define some `int order(recovery_stage)` function that is not persistent to define the ordering in which we the stages get applied.

andrwng replied:
Regarding safety of the current code, I shared this with Andrew over a DM (TL;DR the current code isn't safe; we should either change how the order of `actions` is computed to be decoupled from the enum numerical value, or lump roles recovery in with ACLs, as mentioned elsewhere):

> there's some nuance here: it's true that the first time we run through recovery, the ordering of the returned actions will be correct in that the order will be recovered_acls -> recovered_roles -> recovered_topics -> ..., but we need to be mindful that recovery can be stopped on one node and restarted on another node if there's a leadership change of the controller
>
> when that happens, a new leader will re-evaluate the work that needs to be done, and recompute actions. Let's say hypothetically the first leader recovered up to recovered_roles meaning its cur_stage is recovered_roles (11), and then changed leadership. When the new leader calls get_actions() , it will think that is current stage is past even recovered_topic_data (7), and so the new leader will skip recovering topic data

pgellert reviewed (commented):
Looks pretty good. Can you please update the cover letter as well? (add a description, tick "none - papercut/not impactful enough to backport" if we don't end up backporting; remove the `* none` bullet point)

nguyen-andrew commented:
Force push:
- Moved role recovery within acl recovery stage instead of its own stage.
- Update role reconciliation to examine the role store. It'll only create roles from the snapshot that are missing in the store, and it'll also add any members that are missing from existing roles.
- Added more tests to validate role reconciliation and recovery.

nguyen-andrew commented:
Force push to fix python type checking errors.

nguyen-andrew committed:
chore: update type checking strictness

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74856-->
<!--bc13099e-deef-4ea6-b9b5-8b01c1bc2d9e-->
<!--a520e08b-057d-461a-84cf-6bfe4fa71407-->
<!--ac6b1fc3-a056-46b6-9047-c991880ac96b-->
<!--7b2ae9ff-f2a0-4d30-abb2-6a240777ae2d-->
<!--title-->
## Retry command for [Build#74856](https://buildkite.com/redpanda/redpanda/builds/74856)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
tests/rptest/tests/tls_version_test.py::TLSVersionTestECDSA.test_change_version@{"version":3}
```
<!--/body-->

IoannisRP commented on src/v/cluster/cloud_metadata/cluster_recovery_backend.cc:
```diff
@@ -225,6 +228,32 @@ ss::future<cluster::errc> cluster_recovery_backend::do_action(
+                  roles_retry.get_deadline());
+                break;
+            default:
+                vunreachable("Invalid role action");
```
i don't think this is needed. If someone were to add an extra value in the enumerator, this would make it a runtime error. Without this, it would be a compile time error.

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/4dfe98b5933175f8f37916ef79d43f38416fe559..695e6a2a0c92cb642cd81c290feabf026dffa0b1) to address [pr comment](https://github.com/redpanda-data/redpanda/pull/28166#discussion_r2459808252).

pgellert commented on src/v/cluster/cluster_recovery_reconciler.cc:
```diff
@@ -99,7 +99,39 @@ controller_snapshot_reconciler::get_actions(
+                actions.roles.emplace_back(
+                  cluster::role_recovery::action::update,
+                  snap_role.name,
+                  security::role{std::move(merged_members)});
```
This is a product question, but should we merge or skip roles that already exist?

@andrwng do you have a strong opinion on this based on general WCR requirements?

cc @michael-redpanda

andrwng replied:
I don't have a strong intuition about this, so here's a question: if roles already exists in the new cluster, could merging new things to an existing role cause adverse effects to ongoing workloads on the new cluster? If not, merging seems worthwhile.

pgellert replied:
Closing the loop here that after talking to product, we've decided to skip roles that already exist.

pgellert commented on src/v/cluster/cloud_metadata/cluster_recovery_backend.cc:
```diff
@@ -58,6 +58,7 @@ cluster_recovery_backend::cluster_recovery_backend(
   cluster::members_table& members_table,
   features::feature_table& features,
   security::credential_store& creds,
+  security::role_store& roles,
```
Let's squash this commit into the first one. When introducing new code, we prefer to include its complete behavior in a single commit rather than spreading it across multiple commits. This makes it easier to review and understand the code's impact without having to track how its behavior evolves across different commits.

If you wanted to reduce the commit size, you could have a pre-refactor commit that wires in the role_store into cluster_recovery_backend and leaves it still unused. But I actually don't think that's necessary here.

pgellert commented on src/v/cluster/cloud_metadata/tests/cluster_recovery_backend_test.cc:
```diff
@@ -570,3 +570,118 @@ TEST_F(ClusterRecoveryBackendTest, TestRecoverFailedDownload) {
+          expected_members.size(),
+          role_members.size());
+        ASSERT_EQ(role_members, expected_members);
+    }
```
Nice test! 🔥

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/695e6a2a0c92cb642cd81c290feabf026dffa0b1..9229fc57a4ed3603d860f1a957dfad0132d5f721):
- Removed merging of roles. Will simply skip over existing roles to be consistent with user and ACL recovery since the expectation is that the cluster should by empty when WCR is ran. Also removed test that exercised this role merging logic.
- Updated test to use a set instead of a Counter.
- Clean up of commits

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28166-v24.3.x-974 remotes/upstream/v24.3.x
git cherry-pick -x 1746897ee6 753ef2c45c 7b9eaba265 9229fc57a4
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18912855802).

vbotbuildovich commented:
Failed to create a backport PR to v25.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28166-v25.1.x-447 remotes/upstream/v25.1.x
git cherry-pick -x 1746897ee6 753ef2c45c 7b9eaba265 9229fc57a4
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18912858677).

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28166-v25.2.x-510 remotes/upstream/v25.2.x
git cherry-pick -x 1746897ee6 753ef2c45c 7b9eaba265 9229fc57a4
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18912855835).
